### ISSUE TITLE
Use of GPU delegate increases inference runtime in Android's Native environment (C++) but not using Java API

### ISSUE THREAD

**Issue by muchorange** [tensorflow/tensorflow#42709]
**System information**

- Have I written custom code: I did
- OS Platform and Distribution: Linux Ubuntu 16.04
- Mobile device: Samsung Galaxy S10
- TensorFlow installed from: binary
- TensorFlow version: v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: 3.6.9

**Describe the current behavior**

I'm trying to run a simple UNet model that runs on Android's Native environment and uses the GPU delegate.
I'm testing the same model on `Float32` / `Uint8` fully-quantized (same model, separate `.tflite` files) and also with / without GPU delegate on both cases (so 4 different types of inference in total).

- When testing with / without the GPU delegate - inference runtimes are the same.
- `ModifyGraphWithDelegate()` returns `kTfLiteOk`, but GPU delegate is not applied, and returns an error:
`Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors.`
- Model (see below) shouldn't have any dynamic sized tensors.
- When checked the model input Dims `(B, H, W, C)`, I got an unexpected result.

Tensorflow model was built with `(1, 128, 128, 12)`, but with Android and Tensorflow-Lite I get `(4, 1, 128, 128)`.

**Describe the expected behavior**

Tensorflow model built with `(1, 128, 128, 12)`, should result in Tensorflow-Lite's `(1, 128, 128, 12)` model, and GPU delegate should be applied and reduce inference time on `Float32` type inference.

**Standalone code to reproduce the issue**

Sample C++ code which runs in Android Studio's NDK (bare code, excluding status and sanity checks):
```
std::unique_ptr<tflite::FlatBufferModel> model;
std::unique_ptr<tflite::Interpreter> interpreter;
model = tflite::FlatBufferModel::BuildFromFile(MODEL_FILENAME_HERE.c_str());
model -> error_reporter();
tflite::ops::builtin::BuiltinOpResolver resolver;
tflite::InterpreterBuilder(*model, resolver)(&interpreter);
TfLiteGpuDelegateOptionsV2 gpu_opts = {
        .is_precision_loss_allowed = -1,
        .inference_preference =
                TFLITE_GPU_INFERENCE_PREFERENCE_FAST_SINGLE_ANSWER,
        .inference_priority1 = TFLITE_GPU_INFERENCE_PRIORITY_MAX_PRECISION,
        .inference_priority2 = TFLITE_GPU_INFERENCE_PRIORITY_AUTO,
        .inference_priority3 = TFLITE_GPU_INFERENCE_PRIORITY_AUTO,
        .experimental_flags = TFLITE_GPU_EXPERIMENTAL_FLAGS_NONE,
        .max_delegated_partitions = 1,
};
gpu_opts.inference_preference =
        TFLITE_GPU_INFERENCE_PREFERENCE_SUSTAINED_SPEED;
gpu_opts.inference_priority1 =
        conf.allow_fp16 ? TFLITE_GPU_INFERENCE_PRIORITY_MIN_LATENCY
                : TFLITE_GPU_INFERENCE_PRIORITY_MAX_PRECISION;
auto * gpu_delegte = TfLiteGpuDelegateV2Create(&gpu_opts);
if (USE_GPU_DELEGATE_FLAG) {
        interpreter->ModifyGraphWithDelegate(gpu_delegte);
}
interpreter->AllocateTensors();
int input_tensor_index = interpreter -> inputs()[0];
TfLiteTensor * input_tensor = interpreter -> tensor(input_tensor_index);

// checkpoint - after here there's more code to set input, invoke interpreter, get inference output, etc..

...
```

At `checkpoint`, get input dims by logging them or any other form for example:
```
int d1 = input_tensor -> dims[0];
int d2 = input_tensor -> dims[1];
int d3 = input_tensor -> dims[2];
int d4 = input_tensor -> dims[3];
```

**Other info / logs** Include any logs or source code that would be helpful to

Simple UNet `.tflite` model files [can be found here](https://gofile.io/d/PVJiBY).

When logging with:
```
"Input dims are %dx%dx%dx%d",
                input_tensor -> dims[0],
                input_tensor -> dims[1],
                input_tensor -> dims[2],
                input_tensor -> dims[3]
```
Log shows `Input dims are 4x1x128x128`.

Running inference without GPU delegate shows nothing special in log, but running with GPU shows:
`Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors.`

**Comment by orangesomethingorange** [comment]
I trimmed down the model to a single-operator (`Conv2D`) model ([model file here](https://gofile.io/d/R87ROS)).
The `Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors` error doesn't occur on the single-operator model.
The GPU delegate loads successfully, though I didn't use any 'dynamic' operators in the full Simple UNet model - the full model only has `Conv2D`, `MaxPooling2D` and `UpSampling2D`.

**Comment by orangesomethingorange** [comment]
I replaced `UpSampling2D` with `Resize` and now it loads the GPU delegate successfully ([model files here](https://gofile.io/d/jw7rSe)), but runtimes are exceptionally long with the GPU delegate running inference times 2x-6x longer (default GPU delegate config) compared to no-delegate inference on F32 (non-quantized) models.

**Comment by orangesomethingorange** [comment]
I don't see any OpenCL / OpenGL issues as described in https://github.com/tensorflow/tensorflow/issues/41769

**Comment by impjdi** [comment]
The error message you quoted is coming from `subgraph.cc`, you can do a quick grep to find out.  Anyway, that means some of the tensor dimensions are not fixed, e.g. the batch dimension or the height / width dimensions.  Those are not supported by the GPU and all tensor dimensions need to be specified without any unknowns.

**Comment by orangesomethingorange** [comment]
> The error message you quoted is coming from `subgraph.cc`, you can do a quick grep to find out. Anyway, that means some of the tensor dimensions are not fixed, e.g. the batch dimension or the height / width dimensions. Those are not supported by the GPU and all tensor dimensions need to be specified without any unknowns.

Thanks @impjdi ,

Please note that replacing `UpSampling2D` solved it (I've included the model files [here](https://github.com/tensorflow/tensorflow/issues/42709#issuecomment-683413400)).

The issue is that running the same model with the Java API and the C++ API doesn't show the same behavior:
Using the GPU delegate via the **C++ API** (in native) on a `Float32` model **increases** the inference runtime, while doing so with **Java API** (same model file, same input) **decreases** inference runtime.

**Comment by orangesomethingorange** [comment]
The extra runtime of the GPU delegate in the C++ API comes from here:
`interpreter -> ModifyGraphWithDelegate(gpu_delegte)`

There's no extra runtime when using the Java API.
On the Java API, inference runtimes get significantly shorter (compared to CPU) when using the GPU delegate (on `Float32`).

Here's basically how I use the C++ API:

```
// setup interpreter
std::unique_ptr<tflite::FlatBufferModel> model;
std::unique_ptr<tflite::Interpreter> interpreter;
model = tflite::FlatBufferModel::BuildFromFile(MODEL_NAME);
tflite::ops::builtin::BuiltinOpResolver resolver;
tflite::InterpreterBuilder(*model, resolver)(&interpreter);
interpreter -> UseNNAPI(false);
interpreter -> SetAllowFp16PrecisionForFp32(false);
interpreter -> SetNumThreads(NUMBER_OF_THREADS);

// setup and create GPU delegate
TfLiteGpuDelegateOptionsV2 gpu_opts = {
        .is_precision_loss_allowed = -1,
        .inference_preference =
                TFLITE_GPU_INFERENCE_PREFERENCE_SUSTAINED_SPEED,
        .inference_priority1 = TFLITE_GPU_INFERENCE_PRIORITY_MAX_PRECISION,
        .inference_priority2 = TFLITE_GPU_INFERENCE_PRIORITY_AUTO,
        .inference_priority3 = TFLITE_GPU_INFERENCE_PRIORITY_AUTO,
        .experimental_flags = TFLITE_GPU_EXPERIMENTAL_FLAGS_ENABLE_QUANT,
        .max_delegated_partitions = 1,
};
auto * gpu_delegte = TfLiteGpuDelegateV2Create(&gpu_opts);

// apply GPU delegate
interpreter -> ModifyGraphWithDelegate(gpu_delegte);

// allocate tensors, invoke interpreter, etc...
```

Am I misusing the GPU delegate instance?

Calling `ModifyGraphWithDelegate` can take 10x longer than calling `interpreter -> Invoke()`.
For example (using the models I shared) on a 0.2Mp input:
- **Without the GPU delegate**: calling `interpreter -> Invoke()` takes ~2500 [ms].
- **With the GPU delegate**: calling `interpreter -> Invoke()` takes ~650 [ms], but calling `ModifyGraphWithDelegate` takes ~7000 [ms].

**Comment by impjdi** [comment]
Yeah, OpenCL does an exhaustive tuning for the fastest possible inference when you specify `TFLITE_GPU_INFERENCE_PREFERENCE_SUSTAINED_SPEED`.  If faster initialization is of your concern, use `TFLITE_GPU_INFERENCE_PREFERENCE_FAST_SINGLE_ANSWER`; inference may not be as optimal but may give you a reasonable init time.

**Comment by orangesomethingorange** [comment]
> Yeah, OpenCL does an exhaustive tuning for the fastest possible inference when you specify `TFLITE_GPU_INFERENCE_PREFERENCE_SUSTAINED_SPEED`. If faster initialization is of your concern, use `TFLITE_GPU_INFERENCE_PREFERENCE_FAST_SINGLE_ANSWER`; inference may not be as optimal but may give you a reasonable init time.

Thanks @impjdi ,

I've tried playing around with these flags - this still doesn't explain how the Java API performs much faster comparing to the C++ API, when Java is just a wrapper for the same C++ code (which I assume initializes the GPU delegate with default settings).

### Example test

- Model (see [here](https://github.com/tensorflow/tensorflow/issues/42709#issuecomment-683413400)) is `Float32` datatype
- Input (for this test) is a 0.2 Mp image (any image will do, though)
- I'm running an inference in 2 Android apps - one has Java API and the other has C++ API.
The test includes providing the same input, to the same model, in all test cases.

### Results

- 0.2 Mp input / `Float32` model / Java API:
Total inference runtime - with no delegate: **4057 [ms]**
Total inference runtime* - with GPU delegate: **541 [ms]**

- 0.2 Mp input / `Float32` model / C++ API / `TFLITE_GPU_INFERENCE_PREFERENCE_FAST_SINGLE_ANSWER` flag:
Total inference runtime - with no delegate: **2439 [ms]**
Total inference runtime - with GPU delegate: **3608 [ms] (843 [ms])**
Net inference runtime** - with GPU delegate: **843 [ms]**

- 0.2 Mp input / `Float32` model / C++ API / `TFLITE_GPU_INFERENCE_PREFERENCE_SUSTAINED_SPEED` flag:
Total inference runtime - with no delegate: **2420 [ms]**
Total inference runtime - with GPU delegate: **7035 [ms]**
Net inference runtime** - with GPU delegate: **680 [ms]**

\* _Total inference runtime_ with Java API includes the `ModifyGraphWithDelegate` call runtime
\*\* _Net inference runtime_ with C++ API refers to the runtime without `ModifyGraphWithDelegate` runtime

### Discussion

- In the case without selecting any delegate, is the inference runtime lower with the C++ API since it uses NNAPI delegate by default?
- When running the inference via Java API, runtime is significantly improved when selecting the GPU delegate, compared to no delegate (more than 7x faster!)
- The runtime when using Java API includes the call to `ModifyGraphWithDelegate`, and is still significantly faster.
- When running the (same) inference via C++ API, runtime isn't improved, and is in fact worse, when selecting the GPU delegate, compared to no delegate (1.5x - 3x slower!)
- Measuring the net inference runtime separately reveals that `ModifyGraphWithDelegate` takes the lion share of the runtime.
- Changing the `inference_preference` flags in the C++ API did affect `ModifyGraphWithDelegate`, but inference runtime was still worse than Java API.

How is it possible that in Java API the call to `ModifyGraphWithDelegate` barely takes up any runtime?

**Comment by impjdi** [comment]
Note that `ModifyGraphWithDelegate` is supposed to be called only once at initialization time and `Invoke` multiple times.  So, you can't really say "total inference runtime".   Anyway...

There is a possibility that when using Java, you're using OpenGL, and thus you see a much shorter initialization.  Can you check which backend (OpenGL vs OpenCL) is being triggered when using Java's GPU delegate?

**Comment by orangesomethingorange** [comment]
On the C++ API app, I manually call `ModifyGraphWithDelegate` only once.
On the Java API app, the Java wrapper is doing the same - and `ModifyGraphWithDelegate` is called only once.

** All the following inferences were performed with the same input, same model file, etc.

Running inference with GPU delegate via C++ API, log contains the following:
```
...
15:18:50.182  I/OpenGLRenderer: Davey! duration=984ms; Flags=1, IntendedVsync=5318679868849, Vsync=5318679868849, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=5318685995521, AnimationStart=5318686085313, PerformTraversalsStart=5318686104167, DrawStart=5319563188854, SyncQueued=5319595133490, SyncStart=5319595886562, IssueDrawCommandsStart=5319595981667, SwapBuffers=5319660922500, FrameCompleted=5319664853489, DequeueBufferDuration=135000, QueueBufferDuration=3184000, 
15:18:50.223  I/OpenGLRenderer: Davey! duration=987ms; Flags=0, IntendedVsync=5318717434628, Vsync=5319684101256, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=5319691200364, AnimationStart=5319691309010, PerformTraversalsStart=5319692114219, DrawStart=5319697963906, SyncQueued=5319700952031, SyncStart=5319702011094, IssueDrawCommandsStart=5319702115885, SwapBuffers=5319704500677, FrameCompleted=5319705990573, DequeueBufferDuration=336000, QueueBufferDuration=881000, 
15:19:27.618  I/OpenGLRenderer: Davey! duration=1193ms; Flags=2, IntendedVsync=5355908480170, Vsync=5355908480170, OldestInputEvent=0, NewestInputEvent=0, HandleInputStart=5355908480170, AnimationStart=5355908480170, PerformTraversalsStart=5355908480170, DrawStart=5355908480170, SyncQueued=5355916279882, SyncStart=5355916281496, IssueDrawCommandsStart=5355916887746, SwapBuffers=5356711030350, FrameCompleted=5357101705559, DequeueBufferDuration=344382000, QueueBufferDuration=390205000, 
>>> Inference Invoked Here <<<
15:19:32.689  I/tflite: Initialized OpenCL-based API.
15:19:33.682  I/OpenGLRenderer: Davey! duration=7391ms; Flags=0, IntendedVsync=5355773403908, Vsync=5363056736950, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=5363072122171, AnimationStart=5363072310817, PerformTraversalsStart=5363073168473, DrawStart=5363115731181, SyncQueued=5363122091129, SyncStart=5363123004619, IssueDrawCommandsStart=5363157759879, SwapBuffers=5363163700556, FrameCompleted=5363165528265, DequeueBufferDuration=976000, QueueBufferDuration=729000, 
```

- NOTE: GPU delegate options are default options, with `experimental_flags` set to `TFLITE_GPU_EXPERIMENTAL_FLAGS_ENABLE_QUANT` (though it's not used - model is `Float32`).


Running inference with GPU delegate via Java API, log contains the following:
```
...
>>> Inference Invoked Here <<<
15:24:28.558  I/tflite: Initialized OpenCL-based API.
15:24:29.962  I/OpenGLRenderer: Davey! duration=1862ms; Flags=0, IntendedVsync=5657581696841, Vsync=5659381696769, OldestInputEvent=9223372036854775807, NewestInputEvent=0, HandleInputStart=5659383255964, AnimationStart=5659383328516, PerformTraversalsStart=5659384314922, DrawStart=5659417704141, SyncQueued=5659423333829, SyncStart=5659424487579, IssueDrawCommandsStart=5659434430547, SwapBuffers=5659441704349, FrameCompleted=5659445346954, DequeueBufferDuration=697000, QueueBufferDuration=2316000, 
```

1. I tried to change delegate options on C++ API to include `TFLITE_GPU_EXPERIMENTAL_FLAGS_CL_ONLY` or `TFLITE_GPU_EXPERIMENTAL_FLAGS_GL_ONLY` but output remain the same - no other OpenCL / OpenGL messages on log other than those seen above
2. HW supports `OpenGL ES 3.2` and `OpenCL 2.0`
3. From your comment here: https://github.com/tensorflow/tensorflow/issues/32840#issuecomment-560536940, I guess that in both cases OpenCL is used.
4. Note `OpenGL duration`s: `7391ms` for the C++ and `1862ms` for the Java.
5. Setting `TFLITE_GPU_INFERENCE_PREFERENCE_FAST_SINGLE_ANSWER` does help (duration is `3871ms`), but still Java duration is considerably lower, every time.

Basically I'm just trying to run an inference via the C++ API, in the same way the Java API does.
All settings are just basic default options - what could possibly explain the higher performance the Java API have?

**Comment by ShifengJin** [comment]
How read the output data? 
when I call
TfLiteTensor* predict_tensor = Interpreter->tensor(Interpreter->outputs()[1]);
output = predict_tensor->data.f;
the output data is Abnormal。

**Comment by impjdi** [comment]
@orangesomethingorange Sorry for the slow reply here; have missed this update of yours.  I can't think of anything that would make Java execution faster than the C++ build.  Maybe something related to compilation flags?  Are you doing `bazel build --config android_arm64 -c opt` ?

@ShifengJin `[0]` instead of `[1]` maybe?
